crates/sui-json-rpc/src/authority_state.rs

```diff
@@ -0,0 +1,654 @@
+    ) -> Result<Object, Self::Error> {
+        Ok(self.get_past_object_read(id, *version)?.into_object()?)
+    }
+
+    async fn find_object_lt_or_eq_version(
+        &self,
+        id: &ObjectID,
+        version: &SequenceNumber,
+    ) -> Result<Option<Object>, Self::Error> {
+        let database = self.get_db();
+        let id = *id;
+        let version = *version;
+        spawn_monitored_task!(async move { database.find_object_lt_or_eq_version(id, version) })
+            .await
+            .map_err(StateReadError::from)
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum StateReadInternalError {
+    #[error(transparent)]
+    SuiError(#[from] SuiError),
+    #[error(transparent)]
+    JoinError(#[from] JoinError),
+    #[error(transparent)]
+    Anyhow(#[from] anyhow::Error),
+}
+
+#[derive(Debug, Error)]
+pub enum StateReadClientError {
+    #[error(transparent)]
+    SuiError(#[from] SuiError),
+    #[error(transparent)]
+    UserInputError(#[from] UserInputError),
+}
+
+/// `StateReadError` is the error type for callers to work with.
+/// It captures all possible errors that can occur while reading state, classifying them into two categories.
+/// Unless `StateReadError` is the final error state before returning to caller, the app may still want error context.
+/// This context is preserved in `Internal` and `Client` variants.
+/// Also implements `From` trait to convert into `RpcError` if this is indeed the final error state.
+#[derive(Debug, Error)]
+pub enum StateReadError {
+    // sui_json_rpc::Error will do the final conversion to generic error message
+    #[error(transparent)]
+    Internal(#[from] StateReadInternalError),
+
+    // Client errors
+    #[error(transparent)]
+    Client(#[from] StateReadClientError),
+}
+
+impl From<SuiError> for StateReadError {
+    fn from(e: SuiError) -> Self {
+        match e {
+            SuiError::IndexStoreNotAvailable
+            | SuiError::TransactionNotFound { .. }
+            | SuiError::UnsupportedFeatureError { .. }
+            | SuiError::UserInputError { .. }
+            | SuiError::WrongMessageVersion { .. } => StateReadError::Client(e.into()),
+            _ => StateReadError::Internal(e.into()),
+        }
+    }
+}
+
+impl From<UserInputError> for StateReadError {
+    fn from(e: UserInputError) -> Self {
+        StateReadError::Client(e.into())
+    }
+}
+
+impl From<JoinError> for StateReadError {
+    fn from(e: JoinError) -> Self {
+        StateReadError::Internal(e.into())
+    }
+}
+
+impl From<anyhow::Error> for StateReadError {
+    fn from(e: anyhow::Error) -> Self {
+        StateReadError::Internal(e.into())
+    }
+}
+
+impl From<StateReadError> for RpcError {
+    fn from(e: StateReadError) -> Self {
+        match e {
+            StateReadError::Client(_) => RpcError::Call(CallError::InvalidParams(e.into())),
+            _ => {
+                let error_object = ErrorObject::owned(-32603, e.to_string(), None::<()>);
```

Comment:
Curious about why our error code is negative

Reply:
It's just the jsonrpc standard